storage/postgres/tree_storage.go

```diff
@@ -31,3 +72,317 @@ func OpenDB(connStr string) (*sql.DB, error) {
+	if err != nil {
+		glog.Warningf("Failed to get merkle subtrees: %s", err)
+		return nil, err
+	}
+	defer rows.Close()
+
+	if rows.Err() != nil {
+		// Nothing from the DB
+		glog.Warningf("Nothing from DB: %s", rows.Err())
+		return nil, rows.Err()
+	}
+
+	ret := make([]*storagepb.SubtreeProto, 0, len(nodeIDs))
+
+	for rows.Next() {
+		var subtreeIDBytes []byte
+		var subtreeRev int64
+		var nodesRaw []byte
+		var subtree storagepb.SubtreeProto
+		if err := rows.Scan(&subtreeIDBytes, &subtreeRev, &nodesRaw); err != nil {
+			glog.Warningf("Failed to scan merkle subtree: %s", err)
+			return nil, err
+		}
+		if err := proto.Unmarshal(nodesRaw, &subtree); err != nil {
+			glog.Warningf("Failed to unmarshal SubtreeProto: %s", err)
+			return nil, err
+		}
+		if subtree.Prefix == nil {
+			subtree.Prefix = []byte{}
+		}
+		ret = append(ret, &subtree)
+
+		if glog.V(4) {
+			glog.Infof("  subtree: NID: %x, prefix: %x, depth: %d",
+				subtreeIDBytes, subtree.Prefix, subtree.Depth)
+			for k, v := range subtree.Leaves {
+				b, err := base64.StdEncoding.DecodeString(k)
+				if err != nil {
+					glog.Errorf("base64.DecodeString(%v): %v", k, err)
+				}
+				glog.Infof("     %x: %x", b, v)
+			}
+		}
+	}
+
+	// The InternalNodes cache is possibly nil here, but the SubtreeCache (which called
+	// this method) will re-populate it.
+	return ret, nil
+}
+
+func (t *treeTX) storeSubtrees(ctx context.Context, subtrees []*storagepb.SubtreeProto) error {
+	if glog.V(4) {
+		glog.Infof("storeSubtrees(")
+		for _, s := range subtrees {
+			glog.Infof("  prefix: %x, depth: %d", s.Prefix, s.Depth)
+			for k, v := range s.Leaves {
+				b, err := base64.StdEncoding.DecodeString(k)
+				if err != nil {
+					glog.Errorf("base64.DecodeString(%v): %v", k, err)
+				}
+				glog.Infof("     %x: %x", b, v)
+			}
+		}
+	}
+	if len(subtrees) == 0 {
+		glog.Warning("attempted to store 0 subtrees...")
+		return nil
+	}
+
+	args := make([]interface{}, 0, len(subtrees))
+
+	for _, s := range subtrees {
+		st := s
+		if st.Prefix == nil {
+			panic(fmt.Errorf("nil prefix on %v", st))
+		}
+		subtreeBytes, err := proto.Marshal(st)
+		if err != nil {
+			return err
+		}
+		args = append(args, t.treeID)
+		args = append(args, st.Prefix)
+		args = append(args, subtreeBytes)
+		args = append(args, t.writeRevision)
+	}
+
+	tmpl, err := t.ts.setSubtreeStmt(ctx, len(subtrees))
+	if err != nil {
+		return err
+	}
+	stx := t.tx.StmtContext(ctx, tmpl)
+	defer stx.Close()
+
+	r, err := stx.ExecContext(ctx, args...)
+	if err != nil {
+		glog.Warningf("Failed to set merkle subtrees: %s", err)
+		return err
+	}
+	_, _ = r.RowsAffected()
+	return nil
+}
+
+func (t *treeTX) Commit() error {
+	if t.writeRevision > -1 {
+		if err := t.subtreeCache.Flush(func(st []*storagepb.SubtreeProto) error {
+			return t.storeSubtrees(context.TODO(), st)
+		}); err != nil {
+			glog.Warningf("TX commit flush error: %v", err)
+			return err
+		}
+	}
+	t.closed = true
```

Comment:
Should the reads/writes to `closed` not be covered by a mutex?

Reply:
Looks like it's the same in the `mysql` version, @Martin2112 do you have thoughts?
The `cloudspanner` implementation by comparison does hold `mu` in `Close()`